File: src/lib/editor/wrap.ts
import { graphemes } from "./grapheme";

export interface WrapMetrics {
  chWidth: number;
  cjkWidth: number;
  tabSize: number;
}

/**
 * Compute the set of column positions where visual rows start, for a soft-wrap
 * at `maxPx` pixels. The first entry is always 0. Subsequent entries are column
 * indices (buffer offsets into the line) where a new wrapped row begins.
 *
 * Word-aware with three-tier priority:
 *   1. Whitespace boundary (highest) — wrap after a space/tab. This is the
 *      natural word break and covers Korean/English/Japanese-with-spaces.
 *   2. CJK boundary (fallback) — any Chinese/Japanese/Korean char boundary is
 *      a legal break when no whitespace opportunity exists in this row. Keeps
 *      pure-CJK text (e.g. a long Japanese sentence without spaces) wrappable.
 *   3. Hard grapheme break (last resort) — an unbroken 100-char URL with no
 *      whitespace or CJK anywhere in the row still needs to wrap somewhere.
 *
 * We prefer (1) over (2) so Korean compound words like "시문집" stay intact
 * when the line has spaces elsewhere. A single grapheme wider than `maxPx`
 * still occupies its own row — we never split a cluster.
 */
export function computeWrapStarts(line: string, maxPx: number, m: WrapMetrics): number[] {
  if (maxPx <= 0 || line.length === 0) return [0];
  const tabPx = m.tabSize * m.chWidth;
  const starts = [0];

  let rowStart = 0;
  let x = 0; // cumulative width from rowStart up to (but not including) the current grapheme
  // Two classes of break opportunity, tracked separately so whitespace wins
  // over CJK-boundary. Korean uses spaces between words — we must prefer
  // wrapping at those over splitting mid-syllable-sequence. Pure CJK text
  // (Chinese/Japanese with no spaces) falls through to the CJK break.
  let wsCol = -1; // last whitespace break in this row (high priority)
  let xAtWs = 0;
  let cjkCol = -1; // last CJK-boundary break in this row (low priority)
  let xAtCjk = 0;
  let prevText: string | null = null;

  for (const g of graphemes(line)) {
    const cw =
      g.text === "\t"
        ? (Math.floor(x / tabPx) + 1) * tabPx - x
        : graphemeCellPx(g.text, m);

    // Classify the boundary just before g. Whitespace boundaries (prev is
    // space/tab) are "word-level" breaks — strongly preferred. CJK boundaries
    // (either side is CJK-like) are "character-level" breaks — used only when
    // no whitespace opportunity exists in this row.
    if (prevText !== null && g.col > rowStart) {
      if (isWs(prevText)) {
        wsCol = g.col;
        xAtWs = x;
      } else if (isCjkLike(prevText) || isCjkLike(g.text)) {
        cjkCol = g.col;
        xAtCjk = x;
      }
    }

    if (x + cw > maxPx && x > 0) {
      if (isWs(g.text)) {
        // Whitespace at the end of a row is allowed to overflow — matches
        // CSS `white-space: normal` semantics. The wrap happens AFTER the
        // whitespace, not before it, so trailing spaces cling to the old row.
        x += cw;
        wsCol = g.col + g.text.length;
        xAtWs = x;
      } else if (wsCol > rowStart) {
        // Prefer word-level wrap.
        starts.push(wsCol);
        rowStart = wsCol;
        x = x - xAtWs + cw;
        wsCol = -1;
        xAtWs = 0;
        cjkCol = -1;
        xAtCjk = 0;
      } else if (cjkCol > rowStart) {
        // No space in this row — fall back to CJK-boundary wrap.
        starts.push(cjkCol);
        rowStart = cjkCol;
        x = x - xAtCjk + cw;
        cjkCol = -1;
        xAtCjk = 0;
      } else {
        // No break opportunity at all — hard wrap at the current grapheme.
        starts.push(g.col);
        rowStart = g.col;
        x = cw;
        wsCol = -1;
        xAtWs = 0;
        cjkCol = -1;
        xAtCjk = 0;
      }
    } else {
      x += cw;
    }

    prevText = g.text;
  }
  return starts;
}

function isWs(g: string): boolean {
  return g === " " || g === "\t";
}

// Same CJK ranges `graphemeCellPx` uses for double-width classification.
// Kept as its own predicate so the wrap logic can ask "is this CJK-like?"
// without measuring the pixel width.
function isCjkLike(g: string): boolean {
  if (g.length > 1) return true;
  const code = g.charCodeAt(0);
  if (code < 0x1100) return false;
  return (
    (code >= 0x1100 && code <= 0x115f) ||
    (code >= 0x2e80 && code <= 0x303e) ||
    (code >= 0x3041 && code <= 0x33ff) ||
    (code >= 0x3400 && code <= 0x4dbf) ||
    (code >= 0x4e00 && code <= 0x9fff) ||
    (code >= 0xa000 && code <= 0xa4cf) ||
    (code >= 0xac00 && code <= 0xd7a3) ||
    (code >= 0xf900 && code <= 0xfaff) ||
    (code >= 0xfe30 && code <= 0xfe4f) ||
    (code >= 0xff00 && code <= 0xff60) ||
    (code >= 0xffe0 && code <= 0xffe6)
  );
}

function graphemeCellPx(g: string, m: WrapMetrics): number {
  if (g.length > 1) return m.cjkWidth;
  const code = g.charCodeAt(0);
  if (code < 0x1100) return m.chWidth;
  if (
    (code >= 0x1100 && code <= 0x115f) ||
    (code >= 0x2e80 && code <= 0x303e) ||
    (code >= 0x3041 && code <= 0x33ff) ||
    (code >= 0x3400 && code <= 0x4dbf) ||
    (code >= 0x4e00 && code <= 0x9fff) ||
    (code >= 0xa000 && code <= 0xa4cf) ||
    (code >= 0xac00 && code <= 0xd7a3) ||
    (code >= 0xf900 && code <= 0xfaff) ||
    (code >= 0xfe30 && code <= 0xfe4f) ||
    (code >= 0xff00 && code <= 0xff60) ||
    (code >= 0xffe0 && code <= 0xffe6)
  ) {
    return m.cjkWidth;
  }
  return m.chWidth;
}

/** Find the sub-row index (within `wrapStarts`) that contains `col`. */
export function subRowAt(wrapStarts: number[], col: number): number {
  let sr = wrapStarts.length - 1;
  while (sr > 0 && wrapStarts[sr] > col) sr--;
  return sr;
}

/**
 * Binary search: find the largest buffer-line index `i` such that
 * `lineYOffset[i] <= visualRow`. `lineYOffset` is a cumulative prefix sum
 * with `lineYOffset[0] === 0` and length `lineCount + 1`.
 */
export function bufferLineAtVisualRow(lineYOffset: number[], visualRow: number): number {
  let lo = 0;
  let hi = lineYOffset.length - 2; // last valid buffer line index
  if (hi < 0) return 0;
  while (lo < hi) {
    const mid = (lo + hi + 1) >> 1;
    if (lineYOffset[mid] <= visualRow) lo = mid;
    else hi = mid - 1;
  }
  return lo;
}

File: src/core/wrap.zig
//! Soft wrap: where a logical line breaks into visual rows.
//!
//! Ported from `src/lib/editor/wrap.ts`. Columns are **byte** offsets into the
//! line, not UTF-16 code units.

const std = @import("std");
const grapheme = @import("grapheme.zig");
const width = @import("width.zig");

pub const Metrics = struct {
    /// Advance of a narrow cell, in pixels.
    ch_width: f64,
    /// Advance of a wide (East Asian) cell, in pixels.
    cjk_width: f64,
    /// Tab stop, in narrow cells.
    tab_size: u8,
};

/// Column positions where visual rows start. Entry 0 is always 0.
pub const Starts = struct {
    cols: []const usize,

    pub fn deinit(self: Starts, gpa: std.mem.Allocator) void {
        gpa.free(self.cols);
    }

    /// Sub-row index containing byte offset `col`.
    ///
    /// A linear backward scan, as in the original: lines have a handful of
    /// sub-rows, and the scan starts from the end where the caret usually is.
    pub fn subRowAt(self: Starts, col: usize) usize {
        var sr = self.cols.len - 1;
        while (sr > 0 and self.cols[sr] > col) sr -= 1;
        return sr;
    }

    /// Byte range of sub-row `sr` within a line of length `line_len`.
    pub fn rowRange(self: Starts, sr: usize, line_len: usize) struct { start: usize, end: usize } {
        const start = self.cols[sr];
        const end = if (sr + 1 < self.cols.len) self.cols[sr + 1] else line_len;
        return .{ .start = start, .end = end };
    }

    pub fn rowCount(self: Starts) usize {
        return self.cols.len;
    }
};

fn isWs(cluster: []const u8) bool {
    return cluster.len == 1 and (cluster[0] == ' ' or cluster[0] == '\t');
}

fn cellPx(cluster: []const u8, m: Metrics) f64 {
    return if (width.clusterIsWide(cluster)) m.cjk_width else m.ch_width;
}

/// Compute where visual rows start for a soft wrap at `max_px`.
///
/// Word-aware with three-tier priority:
///   1. Whitespace boundary (highest) -- the natural word break, and what
///      Korean/English/spaced-Japanese need.
///   2. CJK boundary (fallback) -- any wide-character boundary is a legal break
///      when the row holds no whitespace at all, so an unspaced Chinese or
///      Japanese sentence still wraps.
///   3. Hard grapheme break (last resort) -- a 100-character URL has to wrap
///      somewhere.
///
/// Tier 1 beats tier 2 so Korean compounds like `시문집` stay intact when the
/// line has spaces elsewhere. A cluster wider than `max_px` still gets its own
/// row; clusters are never split.
pub fn computeStarts(
    gpa: std.mem.Allocator,
    line: []const u8,
    max_px: f64,
    m: Metrics,
) !Starts {
    var starts: std.ArrayList(usize) = .empty;
    errdefer starts.deinit(gpa);
    try starts.append(gpa, 0);

    if (max_px <= 0 or line.len == 0) {
        return .{ .cols = try starts.toOwnedSlice(gpa) };
    }

    const tab_px = @as(f64, @floatFromInt(m.tab_size)) * m.ch_width;

    var row_start: usize = 0;
    // Cumulative width from row_start up to (but not including) the current
    // cluster.
    var x: f64 = 0;
    // Two classes of break opportunity, tracked separately so whitespace wins
    // over a CJK boundary.
    var ws_col: ?usize = null;
    var x_at_ws: f64 = 0;
    var cjk_col: ?usize = null;
    var x_at_cjk: f64 = 0;
    var prev: ?[]const u8 = null;

    var it = grapheme.iterate(line);
    while (it.nextCluster()) |g| {
        const is_tab = g.text.len == 1 and g.text[0] == '\t';
        const cw = if (is_tab)
            (@floor(x / tab_px) + 1) * tab_px - x
        else
            cellPx(g.text, m);

        // Classify the boundary just before `g`.
        if (prev) |p| {
            if (g.offset > row_start) {
                if (isWs(p)) {
                    ws_col = g.offset;
                    x_at_ws = x;
                } else if (width.clusterIsCjkLike(p) or width.clusterIsCjkLike(g.text)) {
                    cjk_col = g.offset;
                    x_at_cjk = x;
                }
            }
        }

        if (x + cw > max_px and x > 0) {
            if (isWs(g.text)) {
                // Whitespace at the end of a row may overflow -- CSS
                // `white-space: normal` semantics. The wrap lands *after* it, so
                // trailing spaces cling to the old row.
                x += cw;
                ws_col = g.offset + g.text.len;
                x_at_ws = x;
            } else if (ws_col != null and ws_col.? > row_start) {
                try starts.append(gpa, ws_col.?);
                row_start = ws_col.?;
                x = x - x_at_ws + cw;
                ws_col = null;
                x_at_ws = 0;
                cjk_col = null;
                x_at_cjk = 0;
            } else if (cjk_col != null and cjk_col.? > row_start) {
                try starts.append(gpa, cjk_col.?);
                row_start = cjk_col.?;
                x = x - x_at_cjk + cw;
                cjk_col = null;
                x_at_cjk = 0;
            } else {
                try starts.append(gpa, g.offset);
                row_start = g.offset;
                x = cw;
                ws_col = null;
                x_at_ws = 0;
                cjk_col = null;
                x_at_cjk = 0;
            }
        } else {
            x += cw;
        }

        prev = g.text;
    }

    return .{ .cols = try starts.toOwnedSlice(gpa) };
}

/// Pixel advance from the start of `slice` to byte offset `col` within it.
///
/// Tab stops are measured from the start of the slice, which -- because callers
/// pass a single wrapped sub-row -- means tab stops restart on each visual row.
/// That matches the original (`Editor.svelte:769`).
pub fn advanceTo(slice: []const u8, col: usize, m: Metrics) f64 {
    const tab_px = @as(f64, @floatFromInt(m.tab_size)) * m.ch_width;
    var x: f64 = 0;
    var it = grapheme.iterate(slice);
    while (it.nextCluster()) |g| {
        if (g.offset >= col) break;
        if (g.text.len == 1 and g.text[0] == '\t') {
            x = (@floor(x / tab_px) + 1) * tab_px;
        } else {
            x += cellPx(g.text, m);
        }
    }
    return x;
}

/// Inverse of `advanceTo`: the byte offset nearest to pixel position `px`.
///
/// Rounds to whichever cluster edge is closer, so clicking the right half of a
/// character puts the caret after it.
pub fn colFromPx(slice: []const u8, px: f64, m: Metrics) usize {
    const tab_px = @as(f64, @floatFromInt(m.tab_size)) * m.ch_width;
    var x: f64 = 0;
    var it = grapheme.iterate(slice);
    while (it.nextCluster()) |g| {
        const is_tab = g.text.len == 1 and g.text[0] == '\t';
        const cw = if (is_tab) (@floor(x / tab_px) + 1) * tab_px - x else cellPx(g.text, m);
        if (px < x + cw) {
            return if (px - x >= cw / 2) g.offset + g.text.len else g.offset;
        }
        x += cw;
    }
    return slice.len;
}

// -- tests -------------------------------------------------------------------
// Ported from src/lib/editor/wrap.test.ts. Expected columns are byte offsets,
// so the Korean cases differ numerically from the UTF-16 originals (each
// Hangul syllable is 3 bytes rather than 1 unit).

const testing = std.testing;

/// 1px per ASCII cell, 2px per CJK cell -- makes the arithmetic trivial.
const M: Metrics = .{ .ch_width = 1, .cjk_width = 2, .tab_size = 4 };

fn expectStarts(expected: []const usize, line: []const u8, max_px: f64) !void {
    const s = try computeStarts(testing.allocator, line, max_px, M);
    defer s.deinit(testing.allocator);
    testing.expectEqualSlices(usize, expected, s.cols) catch |err| {
        std.debug.print("line={s} max_px={d}\n", .{ line, max_px });
        return err;
    };
}

test "empty line is a single row" {
    try expectStarts(&.{0}, "", 10);
}

test "line that fits is a single row" {
    try expectStarts(&.{0}, "hello", 100);
}

test "breaks after whitespace rather than mid-word" {
    try expectStarts(&.{ 0, 8 }, "foo bar baz", 7);
}

test "breaks at multiple word boundaries" {
    try expectStarts(&.{ 0, 6, 12 }, "aa bb cc dd ee", 5);
}

test "hard-breaks unbroken long strings" {
    try expectStarts(&.{ 0, 3, 6, 9 }, "aaaaaaaaaa", 3);
}

test "wraps between CJK characters with no spaces" {
    // Two 2px syllables per 5px row; byte offsets 0, 6, 12.
    try expectStarts(&.{ 0, 6, 12 }, "안녕하세요", 5);
}

test "prefers whitespace break over mid-CJK break (Korean compound)" {
    // `한국 시문집을`: the space is at byte 6, so the break lands at byte 7 and
    // `시문집을` stays intact instead of splitting mid-compound.
    try expectStarts(&.{ 0, 7 }, "한국 시문집을", 8);
}

test "wraps between ASCII and CJK on the CJK boundary" {
    try expectStarts(&.{ 0, 3 }, "abc안녕", 4);
}

test "a leading grapheme wider than the row still gets its own row" {
    try expectStarts(&.{ 0, 3 }, "안녕", 1);
}

test "trailing whitespace overflows and wraps after the last space" {
    try expectStarts(&.{ 0, 6 }, "foo   bar", 5);
}

test "no break opportunity in the first row falls through to a hard break" {
    try expectStarts(&.{ 0, 3, 6, 9 }, "abcdefgh xy", 3);
}

test "tabs count toward width with tab-stop rounding" {
    try expectStarts(&.{0}, "a\tb", 5);
    try expectStarts(&.{ 0, 2 }, "a\tb", 4);
}

test "subRowAt finds the row containing a column" {
    const s = try computeStarts(testing.allocator, "aa bb cc dd ee", 5, M);
    defer s.deinit(testing.allocator);
    // cols == { 0, 6, 12 }
    try testing.expectEqual(@as(usize, 0), s.subRowAt(0));
    try testing.expectEqual(@as(usize, 0), s.subRowAt(5));
    try testing.expectEqual(@as(usize, 1), s.subRowAt(6));
    try testing.expectEqual(@as(usize, 1), s.subRowAt(11));
    try testing.expectEqual(@as(usize, 2), s.subRowAt(12));
    try testing.expectEqual(@as(usize, 2), s.subRowAt(14));
}

test "advanceTo and colFromPx round-trip on cluster edges" {
    const line = "ab안c";
    try testing.expectEqual(@as(f64, 0), advanceTo(line, 0, M));
    try testing.expectEqual(@as(f64, 2), advanceTo(line, 2, M));
    try testing.expectEqual(@as(f64, 4), advanceTo(line, 5, M)); // after 안
    // Clicking the left half of `안` (x in [2, 3)) lands before it.
    try testing.expectEqual(@as(usize, 2), colFromPx(line, 2.4, M));
    // The right half lands after it.
    try testing.expectEqual(@as(usize, 5), colFromPx(line, 3.6, M));
    // Past the end clamps to the line length.
    try testing.expectEqual(@as(usize, 6), colFromPx(line, 99, M));
}

test "advanceTo restarts tab stops per slice" {
    // A tab at the start of a sub-row advances a full stop.
    try testing.expectEqual(@as(f64, 4), advanceTo("\tx", 1, M));
    // `ab\t` -> the tab fills to the next multiple of 4.
    try testing.expectEqual(@as(f64, 4), advanceTo("ab\tx", 3, M));
}
